### PR TITLE
"/users" API routes are working with GET, POST, PUT:id, DELETE:id, and GET:id

### PR DESCRIPTION
**Pull request by jackbackes**
primary commit is:  a44ca6e

I introduced a module called [sequelize-handlers](https://github.com/botter-workshop/sequelize-handlers) to handle the resource requests to the models. Its methods look like 

```
sequelizeHandlers.create(User)
```

Have you guys thought about what actions will / won't require authentication?

### CODE REVIEW

**Comment by jackbackes** [comment]
the seed file is epic. now it adds up to 100 users, teams, and organizations, with fully completed profiles, and random associations.

**Comment by jackbackes** [comment]
I tested all the routes with postman. there's no mixin logic to them right now (like add a question for a specific assessment, or remove a user from a team, etc), and all the routes are on the root ( ie, api/v1/organizations, api/v1/rubrics, etc ).

File: seed.js [hunk 1]
```diff
@@ -1,57 +1,235 @@
...
+var chalk = require( 'chalk' );
+var db = require( './server/db' );
+var Promise = require( 'sequelize' )
+  .Promise;
+const faker = require( 'faker' );
+const utils = require( './server/app/routes/utils' );
+const {
+  Credentials
+} = utils;
+const Sequelize = require( 'sequelize' );
```
Comment (khumphrey) [inline seed.js]: 👍

File: seed.js [hunk 2]
```diff
@@ -1,57 +1,235 @@
...
+// dependencies
+var chalk = require( 'chalk' );
+var db = require( './server/db' );
+var Promise = require( 'sequelize' )
+  .Promise;
+const faker = require( 'faker' );
+const utils = require( './server/app/routes/utils' );
+const {
+  Credentials
+} = utils;
```
Comment (khumphrey) [inline seed.js:16]: 👍

File: seed.js [hunk 3]
```diff
@@ -1,57 +1,235 @@
...
-            password: 'password'
-        },
-        {
-            email: 'obama@gmail.com',
-            password: 'potus'
-        }
-    ];
+//seed methods
+/** seeds a random number of users 1-100 */
+const seedUsers = function ( n = Math.ceil( Math.random() * 100 ) ) {
```
Comment (khumphrey) [inline seed.js]: this input seems a lot like the function on line 57? Could you reuse the function?

File: seed.js [hunk 4]
```diff
@@ -1,57 +1,235 @@
...
+  let instructors = Array
+    .from( {
+      length: n
```
Comment (khumphrey) [inline seed.js:67]: 👍

File: seed.js [hunk 5]
```diff
@@ -1,57 +1,235 @@
...
+    return User.create( instructorObj );
+  } );
+  return Promise.all( creatingInstructors );
+};
 
-    var creatingUsers = users.map(function (userObj) {
-        return User.create(userObj);
-    });
+/** seeds up to n organizations */
+const seedOrganizations = function ( n = Math.ceil( Math.random() * 100 ) ) {
```
Comment (khumphrey) [inline seed.js]: ^^^^ 
image

File: seed.js [hunk 6]
```diff
@@ -1,57 +1,235 @@
...
+        name
+      }, {
+        role: 'creator'
+      } )
+    } )
+  return Promise.all( organizations );
+}
 
-    return Promise.all(creatingUsers);
+const seedTeams = function ( n = Math.ceil( Math.random() * 100 ) ) {
```
Comment (khumphrey) [inline seed.js]: ^^^ :trollface:

File: seed.js [hunk 7]
```diff
@@ -1,57 +1,235 @@
...
+        teamId
+      } )
+    } )
+  return Promise.all( assessments );
+}
+
+let seedQuestions = function ( n = 20 ) {
+  let questions = Assessment.findAll()
+    .map( assessment => {
+      let N = Math.ceil( Math.random() * n );
```
Comment (khumphrey) [inline seed.js]: randomN 🌀 ❓ ❔

File: seed.js [hunk 8]
```diff
@@ -1,57 +1,235 @@
...
+      let questions = Array.from( {
+        length: N
+      }, question => {
+        let prompt = faker.lorem.words( Math.ceil( Math.random() * 20 ) );
```
Comment (khumphrey) [inline seed.js]: :shipit:

File: seed.js [hunk 9]
```diff
@@ -1,57 +1,235 @@
...
+          assessmentId
+        };
+      } )
+      return Question.bulkCreate( questions );
+    } )
+  return Promise.all( questions );
+}
+
+let seedRubrics = function ( n = 20 ) {
+  console.log( 'seeding rubrics' );
```
Comment (khumphrey) [inline seed.js:174]: Try not to leave consoles in what you want to merge to master (unless you feel it is necessary...)

File: seed.js [hunk 10]
```diff
@@ -1,57 +1,235 @@
...
+  let rubrics = Question.findAll()
+    .map( question => {
+      let N = randomN( n );
+      let rubrics = Array.from( {
+        length: N
+      }, rubric => {
+        let criterion = faker.lorem.words( Math.ceil( Math.random() * 20 ) );
```
Comment (khumphrey) [inline seed.js]: random 💥

File: server/app/index.js [hunk 11]
```diff
@@ -11,7 +11,7 @@ module.exports = function (db) {
 
     // Routes that will be accessed via AJAX should be prepended with
     // /api so they are isolated from our GET /* wildcard.
-    app.use('/api', require('./routes'));
+    app.use('/', require('./routes/'));
```
Comment (khumphrey) [inline server/app/index.js:14]: I can't see the entire file structure, but I wanted to make sure to note that app.use takes './routes/' out of the path moving forward -- this just means that the '/' you might usually include later won't be included. I'm sure you took this into account, but as I can't really see the file structure (this isn't a branch? I'm not sure why not??) I thought I would mention it.

File: server/app/routes/api/v1/annotations/index.js [hunk 12]
```diff
@@ -1,16 +1,23 @@
 'use strict';
```
Comment (khumphrey) [inline server/app/routes/api/v1/annotations/index.js:1]: To me this seems like 2 different PRs -- 1 for the seeding and another for the routes being built out like I'm starting to see below

File: seed.js [hunk 13]
```diff
@@ -1,57 +1,235 @@
...
-    });
+/** seeds up to n organizations */
+const seedOrganizations = function ( n = Math.ceil( Math.random() * 100 ) ) {
+  let organizations = User.findAll( {
+      limit: n
+    } )
+    .map( user => {
+      let name = faker.company.companyName();
+      return user.createOrganization( {
+        name
```
Comment (khumphrey) [inline seed.js:91]: Just as an aside -- this is fine to define variables in the way you have, you can also put the call here on line 89 if you wanted as well. That is more of a personal coding choice on condensed/readable code.

File: server/app/routes/api/v1/annotations/index.js [hunk 14]
```diff
@@ -1,16 +1,23 @@
...
 var router = require('express').Router();
-module.exports = router;
+var _ = require('lodash');
+
+const sequelizeHandlers = require( 'sequelize-handlers' );
+const bluebird = require( 'bluebird' );
+const utils = require('../../../utils');
+const {ensureAuthenticated, ensureIsAdmin, Credentials, respondWith404, _err, db} = utils;
+const Annotation = db().models.annotation;
+const Resource = Annotation;
```
Comment (khumphrey) [inline server/app/routes/api/v1/annotations/index.js:10]: What is the point of having this named twice differently (i.e. Annotation vs Resource)?

File: server/app/routes/api/v1/annotations/index.js [hunk 15]
```diff
@@ -1,16 +1,23 @@
...
+/** see documentation at https://www.npmjs.com/package/sequelize-handlers */
 
-// router.use('/users', require('./users'));
-// router.use('/annotations', require('./annotations'));
-// router.use('/questions', require('.assessments/questions'));
-// router.use('/evaluations', require('.assessments/evaluations'));
-// router.use('/assessments', require('.assessments/assessments'));
+router.get(   '/',  sequelizeHandlers.query(Resource));
+router.get(   '/:id', ensureAuthenticated, sequelizeHandlers.get(Resource));
+router.post(  '/',  ensureAuthenticated,     sequelizeHandlers.create(Resource));
```
Comment (khumphrey) [inline server/app/routes/api/v1/annotations/index.js:16]: I'm not sure if this is accurate, but the spacing looks strange on here. Which brings me to, don't forget to decide on a coding style together so that your project looks cohesive and not like 4 unique people worked on it! <-- as best as you can

File: server/app/routes/api/v1/assessments/assessments.js [hunk 16]
```diff
@@ -1,11 +1,23 @@
 'use strict';
```
Comment (khumphrey) [inline server/app/routes/api/v1/assessments/assessments.js]: Your naming style above was annotations/index.js why is it different for the assessments routes?

File: server/app/routes/api/v1/assessments/assessments.js [hunk 17]
```diff
@@ -1,11 +1,23 @@
...
 var router = require('express').Router();
-module.exports = router;
+var _ = require('lodash');
+
+const sequelizeHandlers = require( 'sequelize-handlers' );
+const bluebird = require( 'bluebird' );
+const utils = require('../../../utils');
+const {ensureAuthenticated, ensureIsAdmin, Credentials, respondWith404, _err, db} = utils;
+const Assessment = db().models.assessment;
+const Resource = Assessment;
```
Comment (khumphrey) [inline server/app/routes/api/v1/assessments/assessments.js:10]: ‼️

File: server/app/routes/api/v1/assessments/evaluations.js [hunk 18]
```diff
@@ -1,11 +1,23 @@
...
 var router = require('express').Router();
-module.exports = router;
+var _ = require('lodash');
+
+const sequelizeHandlers = require( 'sequelize-handlers' );
+const bluebird = require( 'bluebird' );
+const utils = require('../../../utils');
+const {ensureAuthenticated, ensureIsAdmin, Credentials, respondWith404, _err, db} = utils;
+const Evaluation = db().models.evaluation;
+const Resource = Evaluation;
```
Comment (khumphrey) [inline server/app/routes/api/v1/assessments/evaluations.js:10]: 2️⃣

File: server/app/routes/api/v1/annotations/index.js [hunk 19]
```diff
@@ -1,16 +1,23 @@
 'use strict';
 var router = require('express').Router();
-module.exports = router;
+var _ = require('lodash');
+
+const sequelizeHandlers = require( 'sequelize-handlers' );
+const bluebird = require( 'bluebird' );
+const utils = require('../../../utils');
+const {ensureAuthenticated, ensureIsAdmin, Credentials, respondWith404, _err, db} = utils;
```
Comment (khumphrey) [inline server/app/routes/api/v1/annotations/index.js:8]: You don't need to define things you aren't using -- though I am guessing you may use them moving forward since we are at the beginning

File: server/app/routes/api/v1/assessments/rubrics.js [hunk 20]
```diff
@@ -0,0 +1,26 @@
+'use strict';
+var router = require('express').Router();
+var _ = require('lodash');
+
+const sequelizeHandlers = require( 'sequelize-handlers' );
+const bluebird = require( 'bluebird' );
+const utils = require('../../../utils');
+const {ensureAuthenticated, ensureIsAdmin, Credentials, respondWith404, _err, db} = utils;
+const models = db().models;
+const {
```
Comment (khumphrey) [inline server/app/routes/api/v1/assessments/rubrics.js:10]: Why is this one done differently than all of the previous models in the previous routes?

File: server/app/routes/api/v1/index.js [hunk 21]
```diff
@@ -3,10 +3,15 @@ var router = require('express').Router();
...
 
 router.use('/users', require('./users'));
+router.use('/teams', require('./teams'));
+router.use('/organizations', require('./organizations'));
 router.use('/annotations', require('./annotations'));
-router.use('/questions', require('./assessments/questions'));
+router.use('/questions', require('./questions/questions'));
+router.use('/responses', require('./questions/responses'));
+router.use('/templates', require('./templates'));
 router.use('/evaluations', require('./assessments/evaluations'));
```
Comment (khumphrey) [inline server/app/routes/api/v1/index.js:12]: Are evaluations, individual, completed versions of assessments? If so, that seems similar to me to what annotations are to me.

File: server/app/routes/api/v1/index.js [hunk 22]
```diff
@@ -3,10 +3,15 @@ var router = require('express').Router();
 module.exports = router;
 
 router.use('/users', require('./users'));
+router.use('/teams', require('./teams'));
+router.use('/organizations', require('./organizations'));
 router.use('/annotations', require('./annotations'));
-router.use('/questions', require('./assessments/questions'));
+router.use('/questions', require('./questions/questions'));
```
Comment (khumphrey) [inline server/app/routes/api/v1/index.js:9]: Questions seem to be the pieces that make up an assessment, whereas to me responses are what make up an evaluation? 

It seems that you have it so that each question is linked to a user and an assessment so each question has an individual response? If it is that each question has multiple responses that are linked to the question and a user, then to me questions make up assessments and might be better fit inside of that folder. This might be too philosophical though

File: server/app/routes/api/v1/locations/index.js [hunk 23]
```diff
@@ -0,0 +1,23 @@
+'use strict';
+var router = require('express').Router();
+var _ = require('lodash');
+
+const sequelizeHandlers = require( 'sequelize-handlers' );
+const bluebird = require( 'bluebird' );
+const utils = require('../../../utils');
+const {ensureAuthenticated, ensureIsAdmin, Credentials, respondWith404, _err, db} = utils;
+const Location = db().models.location;
+const Resource = Location;
```
Comment (khumphrey) [inline server/app/routes/api/v1/locations/index.js:10]: 🔀

File: server/app/routes/api/v1/organizations/index.js [hunk 24]
```diff
@@ -0,0 +1,37 @@
+'use strict';
+var router = require( 'express' )
+  .Router();
+var _ = require( 'lodash' );
+
+const sequelizeHandlers = require( 'sequelize-handlers' );
+const bluebird = require( 'bluebird' );
+const utils = require( '../../../utils' );
+const {
```
Comment (khumphrey) [inline server/app/routes/api/v1/organizations/index.js:9]: This is visually very different than how you have the other route indices. Be consistent 😃

File: server/app/routes/api/v1/organizations/index.js [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+  db
+} = utils;
+const models = db()
+  .models;
+const {
+  user: User,
+  team: Team,
+  organization: Organization
+} = models;
+const Resource = Organization;
```
Comment (khumphrey) [inline server/app/routes/api/v1/organizations/index.js:24]: I think on line 20-22 you just need {User, Team, Organization} = models. What is the point of user and team though? And same thing with Resource -- why rename this?

File: server/app/routes/api/v1/questions/questions.js [hunk 26]
```diff
@@ -0,0 +1,23 @@
+'use strict';
+var router = require('express').Router();
+var _ = require('lodash');
+
+const sequelizeHandlers = require( 'sequelize-handlers' );
+const bluebird = require( 'bluebird' );
+const utils = require('../../../utils');
+const {ensureAuthenticated, ensureIsAdmin, Credentials, respondWith404, _err, db} = utils;
+const Question = db().models.question;
+const Resource = Question;
```
Comment (khumphrey) [inline server/app/routes/api/v1/questions/questions.js:10]: 👯

File: server/app/routes/api/v1/teams/index.js [hunk 27]
```diff
@@ -0,0 +1,36 @@
+'use strict';
+var router = require( 'express' )
+  .Router();
+var _ = require( 'lodash' );
+
+const sequelizeHandlers = require( 'sequelize-handlers' );
+const bluebird = require( 'bluebird' );
+const utils = require( '../../../utils' );
+const {
```
Comment (khumphrey) [inline server/app/routes/api/v1/teams/index.js:9]: consistency

File: server/app/routes/api/v1/teams/index.js [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+  db
+} = utils;
+const models = db()
+  .models;
+const {
+  user: User,
+  team: Team,
+  organization: Organization
+} = models;
+const Resource = Team;
```
Comment (khumphrey) [inline server/app/routes/api/v1/teams/index.js:24]: re: Organizations/index.js

File: server/app/routes/api/v1/users/index.js [hunk 29]
```diff
@@ -1,27 +1,55 @@
...
+  _err,
+  db
+} = utils;
+const models = db()
+  .models;
+const {
+  user: User,
+  team: Team,
+  organization: Organization
+} = models;
```
Comment (khumphrey) [inline server/app/routes/api/v1/users/index.js:23]: here too

File: server/app/routes/api/v1/users/index.js [hunk 30]
```diff
@@ -1,27 +1,55 @@
...
+      Team, Organization
+    ],
+    attributes: {
+      exclude: [
+        'email', 'password', 'salt'
+      ]
+    }
+  }
+  next();
+} )
```
Comment (khumphrey) [inline server/app/routes/api/v1/users/index.js:38]: 👍

File: server/app/routes/api/v1/users/index.js [hunk 31]
```diff
@@ -1,27 +1,55 @@
...
 
-router.get('/', (req, res, next)=> res.send("happy day!"))
+/** see documentation at https://www.npmjs.com/package/sequelize-handlers */
+router.get( '/', sequelizeHandlers.query( Resource ) );
+router.get( '/:id', ( req, res, next ) => {
+  if ( req.user && req.user.id === req.params.id ) {
+    req.options.attributes.exclude = [ 'salt' ];
```
Comment (khumphrey) [inline server/app/routes/api/v1/users/index.js:44]: This is already being excluded

File: server/app/routes/api/v1/users/spec.js [hunk 32]
```diff
@@ -0,0 +1,35 @@
+'use strict';
```
Comment (khumphrey) [inline server/app/routes/api/v1/users/spec.js:1]: This seems like it should be yet another PR -- one for testing potentially. I'm not going to go over the rest, the following seems to be testing, utils for _everything_ and then models.... Next time just 1 PR guys 😺

File: server/app/routes/index.js [hunk 33]
```diff
@@ -7,6 +7,7 @@ const routerUse = bindRouterToUse(router);
 
 module.exports = router;
 
+
```
Comment (khumphrey) [inline server/app/routes/index.js:10]: this space looks like it shouldn't actually be here

File: server/db/_db.js [hunk 34]
```diff
@@ -7,6 +7,6 @@ global._dbPath = path.join(__dirname, "_db.js");
 var env = require(path.join(__dirname, '../env'));
```
Comment (khumphrey) [inline server/db/_db.js:7]: And now we are adding models into this PR? That should be its own PR. Keep things segregated to keep them clean.

File: server/app/routes/utils.js [hunk 35]
```diff
@@ -1,5 +1,17 @@
 'use strict';
```
Comment (khumphrey) [inline server/app/routes/utils.js:1]: From glancing at this, it seems you put even testing utils in here. I would encourage this to be broken up into testing utils, authentication utils, etc.

Comment (jackbackes) [inline server/app/routes/utils.js:1]: I will refactor.

File: server/app/routes/api/v1/users/index.js (hunk 31 above)
Comment (jackbackes) [inline server/app/routes/api/v1/users/index.js:44]: it has the effect of adding back in the password and email, if the user is logged in.

File: server/app/routes/api/v1/organizations/index.js (hunk 25 above)
Comment (jackbackes) [inline server/app/routes/api/v1/organizations/index.js:24]: the issue with line 20-22 is capitalization. the models object has "user" but not "User"
I had planned to include User and Team on the query, so I included them here.
Resource is the resource for the api for this page.

File: server/app/routes/api/v1/locations/index.js (hunk 23 above)
Comment (jackbackes) [inline server/app/routes/api/v1/locations/index.js:10]: que?

File: seed.js (hunk 7 above)
Comment (jackbackes) [inline seed.js]: ooh! need to refactor!

File: server/app/routes/api/v1/index.js (hunk 21 above)
Comment (jackbackes) [inline server/app/routes/api/v1/index.js:12]: annotations are comments on lines of code

File: server/app/index.js (hunk 11 above)
Comment (jackbackes) [inline server/app/index.js:14]: I'm imagining "routes" as holding all of our routes, rather than only our api routes.

File: server/app/routes/api/v1/annotations/index.js (hunk 12 above)
Comment (jackbackes) [inline server/app/routes/api/v1/annotations/index.js:1]: indeed, I didn't realize that when you commit to the same branch, it updates the pull request. So there are a number of features loaded into one pull request....

File: seed.js (hunk 9 above)
Comment (jackbackes) [inline seed.js:174]: I thought it would be worthwhile to have these in the seed file.

File: server/app/routes/api/v1/annotations/index.js (hunk 15 above)
Comment (jackbackes) [inline server/app/routes/api/v1/annotations/index.js:16]: I had it looking so nice, then made changes and it got all messed up.

**Comment by khumphrey** [comment]
@thejohnbackes @jdhang @jes708 @jancodes I made some comments. Let me know if they don't make sense; if there was a question and you want to answer it and have me see @ comment me!

For next time, I will be checking just **one** PR. This _looked_ like one, but really it should have been at least 4 from what I can see. As such I didn't go through all of it (Omri mentioned that even if it is just one and it is too long that I just can't go through it all). I made notes to where I was and what all I went through. Overall, you guys are making good progress it seems!

File: seed.js (hunk 9 above)
Comment (khumphrey) [inline seed.js:174]: Do you not want them in each stage then? Why just rubrics and like one other if I remember correctly

File: server/app/index.js (hunk 11 above)
Comment (khumphrey) [inline server/app/index.js:14]: Yeah that makes plenty of sense, I was just commenting on the '/' after the word routes on line 14. And also the fact that this PR doesn't seem to be in a branch from master?

File: server/app/routes/api/v1/annotations/index.js (hunk 12 above)
Comment (khumphrey) [inline server/app/routes/api/v1/annotations/index.js:1]: ohhhhhh noted. Yeah it definitely just adds to the PR

File: server/app/routes/api/v1/index.js (hunk 21 above)
Comment (khumphrey) [inline server/app/routes/api/v1/index.js:12]: annotations are comments on lines of code **for** an assessment taken/filled out by an individual right?

File: server/app/routes/api/v1/locations/index.js (hunk 23 above)
Comment (khumphrey) [inline server/app/routes/api/v1/locations/index.js:10]: That was another version of doubling --> why rename location to resource. It can be named resource in the actual function, but they don't have to all be the same thing going in

File: server/app/routes/api/v1/organizations/index.js (hunk 25 above)
Comment (khumphrey) [inline server/app/routes/api/v1/organizations/index.js:24]: Ohhhh gotcha -- this is fine then, though it still seems redundant to rename organization twice

File: server/app/routes/api/v1/users/index.js (hunk 31 above)
Comment (khumphrey) [inline server/app/routes/api/v1/users/index.js:44]: Very nice, then you might want to consider a utility function 'isSelf(id)' that takes an id and determines if that user is the req.user. I feel like you might use that in more places (if not, then no need for a utility function)

**Comment by jackbackes** [comment]
@khumphrey thanks for looking it over! Sorry it was so long! Will definitely try not to do that again.

File: server/app/routes/api/v1/assessments/questions.js [hunk 36]
```diff
@@ -1,11 +0,0 @@
-'use strict';
```
Comment (5617Rls) [inline server/app/routes/api/v1/assessments/questions.js]: Delete sync notes

Comment (5617Rls) [inline server/app/routes/api/v1/assessments/questions.js]: Deleting Router
